File: mask-python/mask_privacy/core/dlp/scorer.py
"""
DLP Confidence Scorer — Proximity-weighted scoring for sensitive data matches.

Combines three independent signals into a single confidence value for each
candidate match, determining whether the JIT encryption hook should fire:

1. **Base risk** — intrinsic leakage probability of the data type.
2. **Proximity boost** — logarithmic-decay bonus for each context keyword
   found near the match within a configurable window.
3. **Validator override** — hard-validator pass forces confidence to 0.99.

The scorer is stateless and thread-safe.
"""

from __future__ import annotations

import math
import logging
from typing import FrozenSet, Optional

_log = logging.getLogger("mask.dlp.scorer")


class DLPConfidenceScorer:
    """Calculate a weighted confidence score for a single regex hit.

    Scoring formula::

        proximity_bonus = Σ  KEYWORD_BOOST / (1 + ln(1 + distance_i))
        raw_score       = base_risk + proximity_bonus
        final_score     = min(MAX_CONFIDENCE, raw_score)

    If a hard-validator returns ``True``, ``final_score`` is overridden
    to ``VALIDATOR_OVERRIDE``.

    Parameters
    ----------
    context_window : int
        Number of characters on each side of the match to scan for keywords.
    keyword_boost : float
        Maximum bonus per keyword (applied at distance = 0).
    validator_override : float
        Confidence assigned when a hard-validator passes.
    max_confidence : float
        Upper clamp for any computed score.
    penalty_factor : float
        Multiplier applied when a hard-validator *fails*.
    """

    DEFAULTS = dict(
        context_window=100,
        keyword_boost=0.10,
        validator_override=0.99,
        max_confidence=0.99,
        penalty_factor=0.65,
    )

    def __init__(self, **overrides) -> None:
        cfg = {**self.DEFAULTS, **overrides}
        self._window: int = cfg["context_window"]
        self._kw_boost: float = cfg["keyword_boost"]
        self._val_override: float = cfg["validator_override"]
        self._ceil: float = cfg["max_confidence"]
        self._penalty: float = cfg["penalty_factor"]

    # ── public API ────────────────────────────────────────────────────────

    def score(
        self,
        *,
        base_risk: float,
        match_start: int,
        match_end: int,
        full_text: str,
        proximity_terms: FrozenSet[str],
        validator_passed: Optional[bool],
    ) -> float:
        """Compute the final confidence for one candidate match.

        Parameters
        ----------
        base_risk
            Intrinsic leakage probability from the pattern descriptor.
        match_start / match_end
            Character offsets of the regex hit inside *full_text*.
        full_text
            The complete input buffer (needed for the context window).
        proximity_terms
            Lowercased keywords to search for near the match.
        validator_passed
            Result from ``DLPValidationEngine.run()``:
            ``True`` → override, ``False`` → penalty, ``None`` → no effect.

        Returns
        -------
        float
            Confidence in ``[0.0, max_confidence]``.
        """
        # Hard-validator short-circuits
        if validator_passed is True:
            return self._val_override
        if validator_passed is False:
            return min(self._ceil, base_risk * self._penalty)

        # Extract the context window around the match
        window_lo = max(0, match_start - self._window)
        window_hi = min(len(full_text), match_end + self._window)
        context_slice = full_text[window_lo:window_hi].lower()

        # Accumulate proximity bonuses
        proximity_bonus = 0.0
        match_mid = (match_start + match_end) / 2.0

        for term in proximity_terms:
            search_pos = 0
            while True:
                idx = context_slice.find(term, search_pos)
                if idx == -1:
                    break
                # Translate window-relative index back to absolute offset
                abs_pos = window_lo + idx
                distance = abs(abs_pos - match_mid)
                bonus = self._kw_boost / (1.0 + math.log(1.0 + distance))
                proximity_bonus += bonus
                search_pos = idx + len(term)

        raw = base_risk + proximity_bonus
        return min(self._ceil, raw)

File: mask-ts/src/core/dlp/scorer.ts
/**
 * DLP Confidence Scorer — Proximity-weighted scoring for sensitive data matches.
 *
 * Combines three independent signals into a single confidence value:
 * 1. Base risk — intrinsic leakage probability of the data type.
 * 2. Proximity boost — logarithmic-decay bonus for each context keyword
 *    found near the match within a configurable window.
 * 3. Validator override — hard-validator pass forces confidence to 0.99.
 *
 * The scorer is stateless and safe for concurrent use.
 */

export interface ScorerConfig {
  contextWindow?: number;
  keywordBoost?: number;
  validatorOverride?: number;
  maxConfidence?: number;
  penaltyFactor?: number;
}

const DEFAULT_CONFIG: Required<ScorerConfig> = {
  contextWindow: 100,
  keywordBoost: 0.10,
  validatorOverride: 0.99,
  maxConfidence: 0.99,
  penaltyFactor: 0.99, // Renamed functionally to validator failure penalty subtraction
};

export interface ScoreInput {
  baseRisk: number;
  matchStart: number;
  matchEnd: number;
  fullText: string;
  proximityTerms: ReadonlySet<string>;
  validatorPassed: boolean | null;
}

/**
 * Calculate a weighted confidence score for a single regex hit.
 *
 * @example
 * ```ts
 * const scorer = new DLPConfidenceScorer();
 * const score = scorer.score({
 *   baseRisk: 0.92,
 *   matchStart: 10,
 *   matchEnd: 21,
 *   fullText: "TC Kimlik No: 10000000146",
 *   proximityTerms: new Set(["kimlik", "tc"]),
 *   validatorPassed: true,
 * });
 * // score === 0.99 (validator override)
 * ```
 */
export class DLPConfidenceScorer {
  private readonly window: number;
  private readonly kwBoost: number;
  private readonly valOverride: number;
  private readonly ceil: number;
  private readonly penalty: number;

  constructor(overrides: ScorerConfig = {}) {
    const cfg = { ...DEFAULT_CONFIG, ...overrides };
    this.window = cfg.contextWindow;
    this.kwBoost = cfg.keywordBoost;
    this.valOverride = cfg.validatorOverride;
    this.ceil = cfg.maxConfidence;
    this.penalty = cfg.penaltyFactor;
  }

  /**
   * Compute the final confidence for one candidate match.
   *
   * @returns Confidence in [0.0, maxConfidence].
   */
  score(input: ScoreInput): number {
    // Hard-validator short-circuits
    if (input.validatorPassed === true) return this.valOverride;
    if (input.validatorPassed === false) {
      return Math.max(0.0, input.baseRisk - this.penalty);
    }

    // Extract the context window around the match
    const windowLo = Math.max(0, input.matchStart - this.window);
    const windowHi = Math.min(input.fullText.length, input.matchEnd + this.window);
    const contextSlice = input.fullText.slice(windowLo, windowHi).toLowerCase();

    // Accumulate proximity bonuses
    let proximityBonus = 0.0;
    const matchMid = (input.matchStart + input.matchEnd) / 2.0;

    for (const term of input.proximityTerms) {
      let searchPos = 0;
      while (true) {
        const idx = contextSlice.indexOf(term, searchPos);
        if (idx === -1) break;
        // Translate window-relative index back to absolute offset
        const absPos = windowLo + idx;
        const distance = Math.abs(absPos - matchMid);
        const bonus = this.kwBoost / (1.0 + Math.log(1.0 + distance));
        proximityBonus += bonus;
        searchPos = idx + term.length;
      }
    }

    const raw = input.baseRisk + proximityBonus;
    return Math.min(this.ceil, raw);
  }
}
